# Jypyter Notebooks and Python Refresh Homework 3
We are going to do the Python Refresh Homework in a juypyter Notebook. In this homework we will review the basic principles of python programming. You will work through a few questions and build some scripts. You will need the Bloom et al Dataset found in this repository. For each question add a cell to put your answer in either as a text cell or a code cell following the question.

### Markdown
Here is a [markdown refresh](https://programminghistorian.org/en/lessons/getting-started-with-markdown)

### General concepts to refresh and lessons to review if you need them:

Here is a nice [Python_Refresh](https://pythonforbiologists.com/introduction) that goes over the majority of the basic python commands.

Remember these Primers from Last Semester

[Python_Primer1](https://github.com/tparchman/BIOL792_course_site/blob/master/week6_pythonI/python_1_primer.md) 

[Lists_for_loops](https://github.com/tparchman/BIOL792_course_site/blob/master/week7_pythonII/python_2_primer.md)

[Input_Output](https://github.com/tparchman/BIOL792_course_site/blob/master/week8_python3/python_3_primer.md)

[Dictionaries](https://github.com/tparchman/BIOL792_course_site/blob/master/week11_python6/primer_python6.md)


### Remember these Common commands as you work through this exercise
	data types
	lists and dictionaries 
	if while and for loops
	reading files



## Part I: Questions

1.	Name one of the basic building blocks of programming and a small definition.

Input and Output are one of the basic building blocks. Input sends information to the programs to be analyzed, and output is what is returned to us- information that was processed and now can be used. 

 _Markdown_ LatTeX: 𝛼2



2.	Match the data type to the appropriate data
    
    **Data type:** string, integer, boolean, floating point
    
    **Data:** true, 123, 'I am a programmer', 34.63221


123 is an integer, 34.63221 is a floating point, true is a boolean, and 'I am a programmer' is a string.

3.	What would I get if I did tried to add these to numbers together ```123``` and ```5```.
		
		a) if the numbers were formatted as strings
		
		b) if they were formatted as integers

In [61]:
###a
# I am creating two sets of integers to be added together
x = 123
y = 5
###b
# I am creating two sets of strings to be added together. Strings are created using quotation marks.
a = "123"
b = "5"
# I am using "print" to show the answers to these additions.
print (x+y)
print (a+b)

128
1235


If the numbers were formatted as a string, the numbers would be combined together as a continuous string. The example shows us that 5 is added to the end of 123 to create 1235, which is not true addition. They are treated as words. With the integers, 123 and 5 are added together in a correct mathematical way, giving us the answer of 128. They are treated as true intergers.

4.	Lists, also called vectors, are one dimensional arrays and are a series of items. From the following list what would the question ```apple_list[3]``` return?  Remember we refer to items in a list by their position. Hint – remember how computers count.
		
		```apple_list=[‘banana’,’pear’,’kumquat’,’pomegranate’,’passion fruit’]```

In [2]:
# I am creating a list of fruits
apple_list = ['banana', 'pear', 'kumquat', 'pomegranate', 'passion fruit']
# I am having python show on the screen the answer to the [3] fruit on the list
print (apple_list[3])

pomegranate


This question is asking is what fruit is in the [3] position. Intuitively, we would think that this is asking us for the 3rd position, or kumquat. But this is not the case. Computers start counting at 0, and so the [3] position is actually the 4th position. The answer is pomegranate.

5.	Name a difference between lists and dictionaries?

Lists are ordered items that are not integers. They are ordered, and always start with 0 as the location of the first item. Dictionaries store paired items, and it is unordered, meaning the order of a dictionary doesn't matter. Each item in a dictionary contains a key, and items are looked up by their keys, making it much easier to look them up than lists.

6.	What would you get from this question from the following dictionary:  ```fruit_dict[‘pear’]```
```		
		fruit_dict = {}
		fruit_dict[‘apple’]=10
		fruit_dict[‘pear’] = 3
		fruit_dict[‘walnut’]=216
```
or in another way


```
fruit_dict = {
    'apple':10,
    'pear':3,
    'walnut':216,
}

print(fruit_dict)
```



In [3]:
#First, I am creating a dictionary with fruits.
#I am leaving the dictionary open "{}" so that I can continue to add more items. 
fruit_dict = {}
#I am adding three fruits or items. Each key has its own value.
fruit_dict["apple"]=10
fruit_dict["pear"] = 3
fruit_dict["walnut"]=216
fruit_dict["watermelon"]=33
#Now I am having python show on the screen the full dictionary- items' value and associated keys.
print(fruit_dict)

#To see a specific item's value, I am having python show to screen the value of "pear".
#I am also adding a label so that I understand what is being shown. 
print("Pear:", fruit_dict["pear"])

{'apple': 10, 'pear': 3, 'walnut': 216, 'watermelon': 33}
Pear: 3


7.	In python what does the function ```.replace()``` do?

.replace() would replace every item that we want to replce in the original string to a new item that we listed for it. 

In [62]:
# Example for .replace
txt = "I like cats"

x = txt.replace("cats", "dogs")

print(x)

I like dogs


8.	How do you print something in python?  For example write a print statement to print “hello world!”

In [4]:
#I am showing to screen the phrase "Hello world!" using the "print" function. 
print ("Hello world!")

Hello world!


9.	Name one place you can go to get help with python?

Our book, Practical Computing for Biologists. Or online resources!

10.	What is an ```if``` statement and when would you use one?

We can think of "if" as a decision maker. It can evalute whether a code that we run is one thing or another, and it can be thought of as a fork in the decision process.

11.	From this list:  ```container_list = [‘can’,’jar’,’hat’]``` What would be returned from this: ```container_list[0:2]```.  Hint: Remember one number is inclusive and one exclusive.

In [5]:
#I am creating a list of items that are containers.
container_list = ["can","jar","hat"]
#I am having it show to screen the output of [0:2].
#This shows the items from position 0 to 1, exluding position 2 and beyond. 
container_list[0:2]

#I am creating another list to practice with. 
#pet_list = ["cat", "dog", "bird", "snake", "guinea pig", "horse"]
#pet_list[3:5]

['can', 'jar']

12.	What is the difference between an ```if``` statement and a ```while``` loop?


An if statement will run though the code finish its execution one it reaches the answer/destination. A while loop will continue until a condition is met. 

13.	What would you get with the command ```fruit_dict.keys()```  after entering this dictionary:  
		fruit_dict = {}
		fruit_dict[‘apple’]=10
		fruit_dict[‘pear’] = 3
		fruit_dict[‘walnut’]=216

In [63]:
fruit_dict.keys()

dict_keys(['apple', 'pear', 'walnut', 'watermelon'])

14.	In this statement: ```InFileName = open(InFileName, ‘r’)``` what does the ‘r’ indicate ?

It indicated that it can be read for the code. It stands for "raw strings".

15.	What does this command do? ```Line.strip(‘\n’)``` why would you use it when printing out lines to the screen?

This removes the line endings. This will help us to read the outfile files in a clearer format.

16.	How would you split a line from a csv file into a list

In [65]:
import csv # I am importing the csv function

f= open('Bloom_etal_2018_Reduced_Dataset.csv') #Opening the file

data = [] #initializing this "data" to put the output into.
for line in f: #creating a for loop
    data_line = line.rstrip().split('\t') #removing the tab delimination, and splitting each line.
    data.append(data_line) #Adding this new additon to data. 
print (data) #Having the output be shown on the screen. 

[['\ufefftaxa,logbodysize,trophic_position,Reg'], ['Alosa_alabamae,1.707570176,0.431363764,diadromous'], ['Alosa_alosa,1.77815125,0.556302501,diadromous'], ['Alosa_fallax,1.77815125,0.556302501,diadromous'], ['Alosa_mediocris,1.77815125,0.612783857,diadromous'], ['Alosa_pseudoharengus,1.602059991,0.544068044,diadromous'], ['Alosa_sapidissima,1.790285164,0.544068044,diadromous'], ['Amazonsprattus_scintilla,0.301029996,0.531478917,non-diadromous'], ['Anchoa_chamensis,0.77815125,0.51851394,non-diadromous'], ['Anchoa_cubana,0.84509804,0.51851394,non-diadromous'], ['Anchoa_delicatissima,0.903089987,0.51851394,non-diadromous'], ['Anchoa_lamprotaenia,0.963787827,0.531478917,non-diadromous'], ['Anchoa_lyolepis,0.84509804,0.531478917,non-diadromous'], ['Anchoa_parva,0.77815125,0.51851394,non-diadromous'], ['Anchoa_schofieldi,0.792391689,0.51851394,non-diadromous'], ['Anchovia_surinamensis,1.093421685,0.477121255,non-diadromous'], ['Anchoviella_balboae,0.857332496,0.505149978,non-diadromous'], [

## Part II:  Practice Scripts
Below each question write the name of the script so I can find them. All scripts should be documented with in-line documentation.

1.	Create a documented python script that would 

		**a.** take this number 112345678911234566 and count the number of 2s in the string and print out the number. 

		**b.** take a sentence from user input, turn it all to lowercase letters and remove the spaces and count the length and print out the length.  You choose the sentence. 


In [6]:
###a.
#I created a string with the numbers provided.
#I made sure to add quotation marks to make it a string.
string1 = ("112345678911234566")

#To count the number of 2s, I had to create a new item "string_count1"
#I then used the .count function on our original string.

string_count1 = string1.count("2")
#I had the program show the answer to the screen.
print("The count of 2 is:", string_count1)

###b.
#I am creating a sentence with capital letters and spaces.
sentence1 = ("Hello my name is Ellen Wilson and I like cats.")

#Now I am removing the capital letters by using the .lower function. 
sentence_lower = sentence1.lower()

#Now I am removing the spaces from the updated sentence, sentence_lower.
#I used the replace function to replace a space with no space.
sentence_spacefree = sentence_lower.replace(" ", "")

#I am printing out this sentence to see if it works. 
print(sentence_spacefree)

#Finally, I am printing the length of sentence_spacefree.
#I am using the function "len" to count all characters in my string.
#I had to be sure to put parantheses around len, otherwise it didn't work. 
print (len(sentence_spacefree))


The count of 2 is: 2
hellomynameisellenwilsonandilikecats.
37


2.	Create a documented python script that will do the following two things. For each task, first write the pseudocode, comment out the pseudocode and beneath the pseudocode write the script.

		**a.** Create a list of numbers (any numbers you like). Then loop through the items in the list adding 1 to every number and print those numbers.

		**b.** Create a dictionary of animals and their sizes (make up whatever you want). Print out the keys of the dictionary. Make a list of all the animals and then write an if else statement to print out the animal name and the word “big” if the weight is over 20 grams and the word “small” if the weight is less than 20 grams. 

Pseudocode A.
First, I will create a list of integers. I will choose to do odd numbers. I will name the list "lista" and the numbers will be a, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21.

Next I will create another integer. This means that I will give another integer outside of my list a value and then I can add this value to my other values. I will call my integer A, and give it a value of 1.

Now I am creating another list, with a formula that will add the newly added A to each number in my list (or x) in my lista. I will call this listb.

Then I will show the contents of listb with the updated addition to each integer to the screen. I will do this by using the print function. I will also add a statement that says "Adding 1:" so that I will be able to understand what I did. 

Now I will do that same process again for pratice. I will create another integer B that will have the value of 5. Then I will add this to lista and create a new list (listc), and print the output to the screen. 


Pseudocode B. 
First, I will initialize an open dictionary. I am calling it wildlife_dict and I will initialize it with open curly brackets. This way, I will be able to add more items to the dictionary. 

Next, I will add 6 items to the dictionary. I will do it by stating the dictionary name, then putting a key in quotation marks inside brackets, and then giving it a value. Ex. "wildlife_dict["elephant"]= 5000000". I will give six varying keys and values. 

Now I am printing just the keys to the screen. I will do this by using the print function, then the dictionary and then .keys after, with an open parentheses. This will print all the keys in my wildlife dictionary. 

Next, I will make a list of all the animals and then write an if else statement to print out the animal name and the word “big” if the weight is over 20 grams and the word “small” if the weight is less than 20 grams.

I will do this by creating a for loop that will go through each key and value in my dictionary. I will have it print to screen the key and whether that is big or small. 

In [35]:
###a.
#I first created a list of odd numbers.
lista = (1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21)

#Next, I had to initialize another integer, which I called A.
#I gave A the value of 1.
A = 1
#I created another list (listb) and gave it the formula listed below.
#x is for any integer in the list, and we are adding each one to A, which has the value of 1.

listb = [x + A for x in lista]
#Now I am printing the results to the screen. We can see that 1 was added to each integar in my list.
print ("Adding 1:", listb)

#I wanted to try this again, but with a different integer to see if it worked.
B = 5
listc = [x + B for x in lista]
print ("Adding 5:", listc)


###b.
#I am creating a wildlife dictionary with animals and thier weight in grams.
#I am initinalizing this dictionary and leaving it open so I can add on to it later.
wildlife_dict = {}
wildlife_dict["elephant"]= 5000000
wildlife_dict["hummingbird"]= 3
wildlife_dict["kangaroo mouse"]= 17
wildlife_dict["capybara"]= 66000
wildlife_dict["rainbow trout"] = 300
wildlife_dict["Pacific tree frog"] = 0.35

print(wildlife_dict)

#I am printing only the keys of the dictionary using the .key function.
#I am having the output be put on the screen.
print(wildlife_dict.keys())


for key, value in wildlife_dict.items(): #I am creating a for loop for the keys and values in my dictionary.
    if value > 20: #If the value is bigger than 20, I will have it print the key and "big".
        print(key, ":", "big")
    else: #If the value is smaller than 20, I will have it print the key and "small".
        print(key, ":", "small")

Adding 1: [2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22]
Adding 5: [6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26]
{'elephant': 5000000, 'hummingbird': 3, 'kangaroo mouse': 17, 'capybara': 66000, 'rainbow trout': 300, 'Pacific tree frog': 0.35}
dict_keys(['elephant', 'hummingbird', 'kangaroo mouse', 'capybara', 'rainbow trout', 'Pacific tree frog'])
elephant : big
hummingbird : small
kangaroo mouse : small
capybara : big
rainbow trout : big
Pacific tree frog : small


3.	Create a documented python script that will open up the file “Bloom_etal_2018_Reduced_Dataset”.  Read through the file and print out the taxon name and their diadromous status. Add up all of the log body sizes and print out the total log body size for all the individuals in the file.  

In [66]:
#First, I am opening the file and adding "r"
f = open("Bloom_etal_2018_Reduced_Dataset.csv", "r")
#I am showing the file content just to see what's inside. 
print(f.read())

taxa,logbodysize,trophic_position,Reg
Alosa_alabamae,1.707570176,0.431363764,diadromous
Alosa_alosa,1.77815125,0.556302501,diadromous
Alosa_fallax,1.77815125,0.556302501,diadromous
Alosa_mediocris,1.77815125,0.612783857,diadromous
Alosa_pseudoharengus,1.602059991,0.544068044,diadromous
Alosa_sapidissima,1.790285164,0.544068044,diadromous
Amazonsprattus_scintilla,0.301029996,0.531478917,non-diadromous
Anchoa_chamensis,0.77815125,0.51851394,non-diadromous
Anchoa_cubana,0.84509804,0.51851394,non-diadromous
Anchoa_delicatissima,0.903089987,0.51851394,non-diadromous
Anchoa_lamprotaenia,0.963787827,0.531478917,non-diadromous
Anchoa_lyolepis,0.84509804,0.531478917,non-diadromous
Anchoa_parva,0.77815125,0.51851394,non-diadromous
Anchoa_schofieldi,0.792391689,0.51851394,non-diadromous
Anchovia_surinamensis,1.093421685,0.477121255,non-diadromous
Anchoviella_balboae,0.857332496,0.505149978,non-diadromous
Anchoviella_brevirostris,0.857332496,0.505149978,non-diadromous
Anchoviella_jamesi,0.6989700

In [68]:
import csv #I am importing the csv function

Bloom = [] #Initializing an open function for the data loop to run through.
f= open('Bloom_etal_2018_Reduced_Dataset.csv', 'r') #opening the file, adding 'r'.
for line in f: #creating a for loop, having it go through each loop
    line = line.strip ('\n') #getting rid of the line endings for an easier read.
    cells = line.split(",") #splitting the items up by the commas, so each will be listed as thier own item.
    Bloom.append( (cells[0], cells[3])) #I am having only column 1 and 4 be added to the output.
print(Bloom) # I am having the output be shown to the screen

[('\ufefftaxa', 'Reg'), ('Alosa_alabamae', 'diadromous'), ('Alosa_alosa', 'diadromous'), ('Alosa_fallax', 'diadromous'), ('Alosa_mediocris', 'diadromous'), ('Alosa_pseudoharengus', 'diadromous'), ('Alosa_sapidissima', 'diadromous'), ('Amazonsprattus_scintilla', 'non-diadromous'), ('Anchoa_chamensis', 'non-diadromous'), ('Anchoa_cubana', 'non-diadromous'), ('Anchoa_delicatissima', 'non-diadromous'), ('Anchoa_lamprotaenia', 'non-diadromous'), ('Anchoa_lyolepis', 'non-diadromous'), ('Anchoa_parva', 'non-diadromous'), ('Anchoa_schofieldi', 'non-diadromous'), ('Anchovia_surinamensis', 'non-diadromous'), ('Anchoviella_balboae', 'non-diadromous'), ('Anchoviella_brevirostris', 'non-diadromous'), ('Anchoviella_jamesi', 'non-diadromous'), ('Anchoviella_lepidentostole', 'diadromous'), ('Anchoviella_manamensis', 'non-diadromous'), ('Anodontostoma_chacunda', 'diadromous'), ('Cetengraulis_edentulus', 'non-diadromous'), ('Chirocentrodon_bleekerianus', 'non-diadromous'), ('Clupeichthys_perakensis', 'n

In [69]:
##I am doing the exact same thing as above,
#but I am only keeping column 2, because that is the information we need to be shown. 

import csv

output = []
f= open('Bloom_etal_2018_Reduced_Dataset.csv', 'r')
for line in f:
    line = line.strip ('\n')
    cells = line.split(",")
    output.append( (cells[1]))
print(output)
#Unfortunately, I do not know how to add these items together right now. 


['logbodysize', '1.707570176', '1.77815125', '1.77815125', '1.77815125', '1.602059991', '1.790285164', '0.301029996', '0.77815125', '0.84509804', '0.903089987', '0.963787827', '0.84509804', '0.77815125', '0.792391689', '1.093421685', '0.857332496', '0.857332496', '0.698970004', '1.064457989', '0.397940009', '1.342422681', '1.113943352', '0.954242509', '0.602059991', '0.903089987', '1.544068044', '0.698970004', '0.903089987', '1.190331698', '1', '1.544068044', '1', '1.387389826', '1', '0.838849091', '0.77815125', '1.113943352', '0.929418926', '1.079181246', '1.505149978', '0.903089987', '0.903089987', '0.672097858', '0.477121255', '0.84509804', '0.973127854', '1', '0.447158031', '1.77815125', '1.041392685']


I am not sure how to add this column together.